### DISCUSSION TITLE
Questions about TypeScript compatibility and Hermes[main] support

### DISCUSSION THREAD

**Discussion by NiuGuohui** [facebook/hermes#1587] [Static Hermes]
I tried out Static Hermes today, and I’m really excited about the performance! However, during the testing process, I had some questions:
It seems like the current version of SH only supports TypeScript at a very basic level. Features like `interface`, `IntersectionType`, Class inheritance, and many other syntax features don't seem to be supported.
I understand that fully supporting all TypeScript syntax is nearly impossible, but I’m wondering about SH's future development. Will it continue to improve TypeScript syntax compatibility (it almost feels like you're building a TypeScript runtime)? Or is there a compatibility solution (such as using a JS compiler to remove highly flexible and unsupported TypeScript syntax before SH compilation)?
Additionally, I’d like to understand whether the development of SH will continue to impact the progress of the current Hermes main version. I noticed that the Hermes version seems to have not updated its language feature support for quite a while, including support for things like Async functions, ES modules, Classes, and method definitions that were mentioned in the documentation some time ago.

**Comment by tmikov** [comment]
Hi, 
Let me address your question about the `main` branch first. It has been in maintenance mode for at least two years now, with only critical bug fixes and API compatibility changes. All development is in the `static_h` branch, which is much more powerful, feature complete, faster, more correct, etc. It is used by Meta internally, it can be used by React Native today with some manual tweaking (Software Mansion has already released a fork that does it), and it will become the default in RN in 2025.

You mentioned ESM. Unfortunately that is a very complicated topic which is out of our hands, because it depends on modifying the existing bundlers. For some more details:
- https://github.com/facebook/hermes/issues/1382#issuecomment-2084413432
- https://github.com/facebook/hermes/issues/1481#issuecomment-2294929931

Your other question was about **TypeScript support**.

We intend to eventually support as much of TypeScript as we can, however it is not our primary focus at the moment, because we are concentrating on completing the basics of the sound type system. When that is finished, we will eventually gradually map more TS features to it. Keep in mind that TypeScript is an extremely complicated language, and a lot of its features (like structural typing) are not compatible with a fast and sound typed implementation.

**Reply by NiuGuohui** [reply]
I understand now! It seems like I may have mixed up the TypeScript type system with SH's sound type system. So, SH’s static compilation optimizations are based on the sound type system, not on any specific type language (TS/Flow), right?
﻿
Can I understand it this way: For future versions that may be officially released, we would actually pre-compile the TS/Flow source code into a form that matches the sound type system syntax before running SH compilation?

**Reply by tmikov** [reply]
> I understand now! It seems like I may have mixed up the TypeScript type system with SH's sound type system. So, SH’s static compilation optimizations are based on the sound type system, not on any specific type language (TS/Flow), right?

Somewhat. We are guided by the existing type systems, Flow and TypeScript - our sound type system is a subset of them. Flow is more sound than TypeScript, for example it relies on *nominal typing* for classes, which is big advantage, so we are somewhat closer to Flow as a result.
﻿
> Can I understand it this way: For future versions that may be officially released, we would actually pre-compile the TS/Flow source code into a form that matches the sound type system syntax before running SH compilation?

Unfortunately such a tool does not exist and would be very difficult to write. If this was possible, Static Hermes would be doing that automatically.

The problem is not the syntax - we are already syntax compatible (more or less), the problem are things that are simply not compatible with an efficient sound type system. They cannot be converted automatically.

Here is a small example to make this clearer:

```js
type MyType = {
  a: number;
  b: string;
  c?: boolean;
};

type OtherType = {
  b: string; // Out of order
  d: string; // Extra field
  a: number; // Out of order
};

function processObject(obj: MyType): void {
  console.log(obj.a, obj.b, obj.c);
}

// Call with an object typed as OtherType
const obj: OtherType = {
  d: "extra",
  b: "hello",
  a: 42,
};

processObject(obj);
```

You see how `MyType` and `OtherType` have dramatically different fields and layouts, and yet, it is valid to call `processObject()` with `OtherType`. We could go into the weeds here, but basically there is no practical way to make this to perform with native speed with ahead-of-time compilation in the context of JavaScript.

So, in order to use Static Hermes typed mode, you would have to modify your code to follow the SH rules, or write new code. The good news is that the SH rules are a *subset* of Flow and TypeScript, so the code remains compatible with them. It is just stricter.

**Reply by NiuGuohui** [reply]
So, based on what you've said, in a real project, even if our code strictly follows SH's syntax, we might still encounter some performance bottlenecks. For example, if we use third-party JS/TS libraries (like lodash) that don’t have an SH-compatible version, this would cause them to fall back to regular JS interpretation.
﻿
The TS/Flow source-to-SH syntax conversion tool I mentioned earlier is actually intended to address such scenarios. By using this tool, we could convert the code in a way that preserves the sound type system’s syntax while removing unsupported TS syntax, so that more code snippets can be optimized during SH compilation.
﻿
Of course, I understand that this may not be the core focus of SH, since the tool I mentioned could probably be implemented as just a Babel plugin. :-)

**Reply by tmikov** [reply]
> that don’t have an SH-compatible version, this would cause them to fall back to regular JS interpretation.

Typed code can't just silently fall back to regular JS. If you are compiling something in typed mode, it has to follow the SH typechecker rules, or there would be a compilation error.  Additionally, typed and untyped code cannot directly share types, because that would break soundness.

> The TS/Flow source-to-SH syntax conversion tool I mentioned earlier is actually intended to address such scenarios. By using this tool, we could convert the code in a way that preserves the sound type system’s syntax while removing unsupported TS syntax, so that more code snippets can be optimized during SH compilation.

Writing such a tool would be more difficult than writing Static Hermes. It would be a significant scientific achievement...

**Reply by NiuGuohui** [reply]
> If you are compiling something in typed mode, it has to follow the SH typechecker rules, or there would be a compilation error.

I successfully executed some regular JavaScript code on shermes. Although the code didn't use the sound type, resulting in not-so-high execution efficiency, I noticed that it seems that SH also supports compile regular JavaScript code.

> Additionally, typed and untyped code cannot directly share types, because that would break soundness.

That is to say, can't SH be compatible with the current JS/TS ecosystem? Does all code have to be rewritten based on the sound type system?

**Reply by tmikov** [reply]
SH is fully compatible with current JS. It supports two modes: **typed** (the new stuff) and **untyped** (good old JavaScript). They can be mixed and can talk to each other, but the sound types, which typed mode relies on, cannot be instantiated directly by untyped code, or that would cause unsoundness.

The idea is that one should use typed mode to optimize hot spots, or to write new code.

**Reply by NiuGuohui** [reply]
> The idea is that one should use typed mode to optimize hot spots, or to write new code.

Alright, I now have a better understanding of the future usage scenarios for SH. I'm looking forward to seeing how SH performs when integrated into projects in the future!
Thank you for taking the time to clarify my doubts.
